### PR TITLE
Remove helper methods on `vdaf::Aggregator`

### PR DESCRIPTION
**Pull request by tgeoghegan**
Remove the methods `Aggregator::{prepare_start, prepare_next,
prepare_finish}`. Using these methods correctly requires the caller to
know statically how many rounds a VDAF requires, because calling
`prepare_next` on a finished preparation or `prepare_finish` on one with
remaining rounds is an error. In order to write a PPM implementation
that is generic over trait `vdaf::Aggregator`, we would have to
introduce a generic means to discover how many rounds a VDAF needs or
whether the `Aggregator::PrepareStep` is on the last round.

The alternative is for a generic implementation to use
`Aggregator::prepare_step`, which returns an enumeration telling the
caller what to do next. Exposing two interfaces for driving the prepare
protocol is more complexity and risk than we need, so we will force
callers to only use one.

This commit also adopts `vdaf::run_vdaf{_prepare}` in `vdaf/poplar1`'s
tests, since the existing tests depended on the helper methods. This
does introduce a signficant performance regression in the unit tests,
because `vdaf::run_vdaf` makes a point of encoding and decoding
messages. With our `ToyIdpf`, this involves encoding and decoding
several hundred thousands of `Field128` values, so `test_poplar1` takes
about 20 seconds on my desktop. Exercising encode/decode of `poplar1`
messages also revealed a bug in `Poplar1InputShare::decode` that this
commit fixes.

### CODE REVIEW

File: src/vdaf.rs [hunk 1]
```diff
@@ -163,64 +164,25 @@ pub trait Aggregator: Vdaf {
...
     ) -> Result<Self::PrepareMessage, VdafError>;
 
     /// Compute the next state transition from the current state and the previous round of input
-    /// messages.
+    /// messages. If this returns [`PrepareTransition::Continue`], then the returned
+    /// [`Self::PrepareMessage`] should be combined with the other aggregator's `PrepareMessage`s from
+    /// this round and passed into another call to this method. This continues until this method
+    /// returns [`PrepareTransition::Finish`], at which point the returned output share may be
+    /// aggregated. If the method returns [`PrepareTransition::Fail`], the aggregator should alert
+    /// its peers as specified in the PPM protocol.
```
Comment (cjpatton) [inline src/vdaf.rs:179]: It's best to not be too specific about the application.

```suggestion
    /// aggregated. If the method returns [`PrepareTransition::Fail`], the aggregator should consider its input share invalid and not attempt to process it any further.
```

File: src/vdaf/poplar1.rs [hunk 2]
```diff
@@ -304,23 +305,28 @@ impl<I: Idpf<2, 2>> Decode<Poplar1VerifyParam> for Poplar1InputShare<I> {
...
+    input_length: usize,
     phantom: PhantomData<I>,
 }
 
 impl<I> Poplar1<I> {
     /// Create an instance of the poplar1 VDAF. The caller provides a cipher suite `suite` used for
-    /// deriving pseudorandom sequences of field elements.
-    pub fn new(suite: Suite) -> Self {
+    /// deriving pseudorandom sequences of field elements, and a input length in bits, corresponding
+    /// to `BITS` as defined in [the `poplar1` VDAF][1].
```
Comment (cjpatton) [inline src/vdaf/poplar1.rs:315]:
```suggestion
    /// to `BITS` as defined in in the [VDAF spec][1].
```

File: src/vdaf/poplar1.rs [hunk 3]
```diff
@@ -822,41 +833,23 @@ mod tests {
 
     #[test]
     fn test_poplar1() {
-        let vdaf: Poplar1<ToyIdpf<Field128>> = Poplar1::new(Suite::Blake3);
-        assert_eq!(vdaf.num_aggregators(), 2);
+        const INPUT_LEN: usize = 16;
```
Comment (cjpatton) [inline src/vdaf/poplar1.rs:836]: to cut down on the runtime of the unit tests, we ought to cut this down to 8 bits. We should be able to modify the tests cases accordingly and still have the same coverage.

**Review by cjpatton — CHANGES_REQUESTED** [review]
Looks great. 20 seconds for unit tests is definitely unacceptable, so let's see if we can't resolve this by reduing the input size for the tests (as discussed inline).

**Review by cjpatton — APPROVED** [review]
LGTM! I totally forgot about the `0b0000_0000` syntax. God, rust is great.

**Comment by tgeoghegan** [comment]
With this change that trims down the input size in the `poplar1` test, the tests run in about half a second on my relatively beefy desktop.

**Pull request merged**